### PR TITLE
[FSDP][Reland] Implement local_state_dict and load_local_state_dict

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73203

1. Implement the framework to allow user to choose among `state_dict`, `local_state_dict`, and `sharded_state_dict`.
2. Implement ShardedTensor compatible local_state_dict() and load_local_state_dict().

Differential Revision: [D34383925](https://our.internmc.facebook.com/intern/diff/D34383925/)

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
LGTM although I'm not sure what particular changes were needed to reland this without breakages in the previous PR.

Looking at the discussion in the original diff, those tests should not have been broken by this PR but they indeed got fixed by the revert (failures are https://github.com/pytorch/pytorch/runs/5260788165?check_suite_focus=true). Added ci all label here to see if they reproduce.

**Comment by rohan-varma** [comment]
Skipping the failing tests for now in https://github.com/pytorch/pytorch/pull/73261
